### PR TITLE
Execute sinfo with batchspawner.run_command

### PR DESCRIPTION
**Pull request by lexming**
We are using `jupyterhub_moss` in our cluster, but we found some trouble with the `sinfo` command. Please consider the following PR which addresses some of this issues.

Changelog:
1. The command `sinfo` is currently executed with `subprocess.check_output()`. However `batchspawner` already provides a more powerful `run_command()` that is asynchronous. This PR removes the dependency on `subprocess` to rely on `batchspawner`.
2. Added the option to easily customize the `sinfo` command
3. Prepend `SlurmSpawner.exec_prefix` to `sinfo`. Anything in `exec_prefix` needed to execute `sbatch` or `squeue`  will also be needed to execute `sinfo`.

### CODE REVIEW

File: jupyterhub_moss/spawner.py [hunk 1]
```diff
@@ -96,12 +101,17 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.options_form = self.create_options_form
 
-    def _get_slurm_info(self):
+    async def _get_slurm_info(self):
```
Comment (t20100) [inline jupyterhub_moss/spawner.py:104]: 👍 

Could you add `batchspawner>=1.0` in `pyproject.toml` for the switch to `async`?

File: jupyterhub_moss/spawner.py [hunk 2]
```diff
@@ -96,12 +101,17 @@ def __init__(self, *args, **kwargs):
...
         """Returns information about partitions from slurm"""
-        # Get number of nodes and idle nodes for all partitions
-        state = check_output(
-            ["sinfo", "-a", "-N", "--noheader", "-o", "%R %t %m"]
-        ).decode("utf-8")
+        subvars = self.get_req_subvars()
+        cmd = " ".join(
+            (
+                format_template(self.exec_prefix, **subvars),
```
Comment (t20100) [inline jupyterhub_moss/spawner.py:109]: 👍 It was missing

File: jupyterhub_moss/spawner.py [hunk 3]
```diff
@@ -297,4 +307,8 @@ def options_from_form(self, formdata: Dict[str, List[str]]) -> Dict[str, str]:
     async def submit_batch_script(self):
         self.log.info(f"Used environment: {self.user_options['environment_path']}")
         self.log.info(f"Used default URL: {self.default_url}")
+
+        # refresh environment to be kept in the job
+        self.req_keepvars = self._req_keepvars_default()
```
Comment (t20100) [inline jupyterhub_moss/spawner.py:312]: What is the motivation to set it here?

It seems not related to the rest of the PR and sounds not always desired (e.g., if `req_keepvars` is set elsewhere).

File: jupyterhub_moss/spawner.py [hunk 4]
```diff
@@ -79,6 +78,12 @@ def _validate_partitions(self, proposal):
                 env.setdefault("add_to_path", True)
         return partitions
 
+    slurm_info_cmd = traitlets.Unicode(
```
Comment (t20100) [inline jupyterhub_moss/spawner.py:81]: I'm wondering if this is flexible enough because the output is tightly coupled with the parsing done in `_get_slurm_info`.

Also, there is maybe a better/simpler way to retrieve this information from SLURM.

**Review by t20100 — COMMENTED** [review]
Hi,

Thanks for the PR!

Using `async` and `run_command()` and prepending `SlurmSpawner.exec_prefix` to `sinfo` were clearly missing.

What was your issue with running `sinfo` beyond the `exec_prefix` not being used?

File: jupyterhub_moss/spawner.py [hunk 5]
```diff
@@ -79,6 +78,12 @@ def _validate_partitions(self, proposal):
...
+        # Get number of nodes and cores for all partitions
+        r"sinfo -a -N --noheader -o \'%R %C %m\'",
```
Comment (t20100) [inline jupyterhub_moss/spawner.py:83]:
```suggestion
        r"sinfo -a -N --noheader -o '%R %t %m'",
```

This is too keep the same command as before and `\'` doesn't seems to work.

File: jupyterhub_moss/spawner.py (hunk 4 above)
Comment (t20100) [reply jupyterhub_moss/spawner.py:81]: I was thinking of something like:
```python
    slurm_info = traitlets.Callable(
        help="""Provides SLURM cluster information.

            It will be called with the Spawner as a single argument and return a mapping: partition name: SlurmInfo.
            It may be async.
    """,
    ).tag(config=True)
```
It leaves more freedom to refactor/extend the default `_get_slurm_info` and also more freedom when this needs to be tuned, but it is at the expense of more code when willing to override it.
Would this work in your case?

File: jupyterhub_moss/spawner.py (hunk 1 above)
Comment (lexming) [reply jupyterhub_moss/spawner.py:104]: good point, done in https://github.com/silx-kit/jupyterhub_moss/pull/64/commits/f1a874dc7f18f1e3a43ce285de1fef5e12cd9497

File: jupyterhub_moss/spawner.py (hunk 4 above)
Comment (lexming) [reply jupyterhub_moss/spawner.py:81]: There might be other ways to get that info from Slurm through its REST API, but that is not default. IMO, since batchspawner uses simple strings to define `batch_submit_cmd` (sbatch) and `batch_query_cmd` (squeue), I would stick to the same format for consistency. Even if somebody wants to use a REST call to get that information, it is very simple to do so through a command.

File: jupyterhub_moss/spawner.py (hunk 5 above)
Comment (lexming) [reply jupyterhub_moss/spawner.py]: you're right, that was an unintended change. We are showing idle cores instead of idle nodes in the table of available resources. Since notebooks do not use full nodes, we think it is more useful for the user to know about cores. If you are interested in this change I will open another PR for it.

**Comment by lexming** [comment]
> What was your issue with running sinfo beyond the exec_prefix not being used?

Hi @t20100 , we have a particular setup where JupyterHub is run in a very minimal container without Slurm. All `s` commands are executed on a remote host through SSH, which adds some latency to the call. That's why using `async` is beneficial and we also need `exec_prefix` with templating as we define the `ssh` command there.

File: jupyterhub_moss/spawner.py (hunk 5 above)
Comment (t20100) [reply jupyterhub_moss/spawner.py]: Indeed, the max number of idle cores available on a node would be more appropriate, PR welcomed!

File: jupyterhub_moss/spawner.py (hunk 3 above)
Comment (t20100) [reply jupyterhub_moss/spawner.py:312]: Thanks for the explanation! Looking at `req_keepvars` in `batchspawner`, it is not meant to be set from the config.

If possible, it would be better to avoid calling private methods, with e.g. [trait_defaults](https://github.com/ipython/traitlets/blob/65a4725a11c776054210583cf9da490e96bd2d23/traitlets/traitlets.py#L1740-L1746):
```suggestion
        # refresh environment to be kept in the job
        self.req_keepvars = self.trait_defaults("req_keepvars")
```

File: jupyterhub_moss/spawner.py (hunk 4 above)
Comment (t20100) [reply jupyterhub_moss/spawner.py:81]: I was thinking of still using `sinfo` from the command line, but to also allow defining the parsing of `sinfo` output because both are tightly coupled, but it is OK like that: this can be added later if needed.

File: jupyterhub_moss/spawner.py (hunk 3 above)
Comment (lexming) [reply jupyterhub_moss/spawner.py]: good point, I was a bit reckless there :slightly_smiling_face:

File: jupyterhub_moss/spawner.py (hunk 4 above)
Comment (lexming) [reply jupyterhub_moss/spawner.py:81]: I see what you mean, tying up the `sinfo` command and the parser in a single option does make sense, but I would do that in a different PR.

**Comment by t20100** [comment]
@lexming if you want that I tag a release with this PR, just let me know.
